### PR TITLE
Refactor kube-cert-agent controllers to use a Deployment.

### PR DESCRIPTION
**Pull request by mattmoyer**
This is a relatively large rewrite of much of the kube-cert-agent controllers. Instead of managing raw Pod objects, they now create a single Deployment and let the builtin k8s controller handle it from there.

This reduces the amount of code we need and should handle a number of edge cases better, especially those where a Pod becomes "wedged" and needs to be recreated.

cc #493 

This is a WIP until I am sure what I want to do about the integration tests. I wrote a very simple test to assert some basic properties of the agent pods.

**Release note**:

```release-note
Fixed a bug where the kube-cert-agent pod could become stuck when cluster nodes were suspended and resumed.
```

### CODE REVIEW

File: internal/controller/kubecertagent/deployer.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+	// If the existing Deployment already matches our desired spec, we're done.
+	if apiequality.Semantic.DeepDerivative(expectedDeployment, existingDeployment) {
+		c.log.Info("existing Deployment is already up to date", "deployment", klog.KObj(expectedDeployment))
+		return nil
+	}
+
+	// Otherwise update the spec of the Deployment to match our desired state and update it.
+	c.log.Info("updating existing Deployment", "deployment", klog.KObj(expectedDeployment))
+	expectedDeployment.Spec.DeepCopyInto(&existingDeployment.Spec)
```
Comment (enj) [inline internal/controller/kubecertagent/deployer.go:146]: You are mutating an informer cache item.

File: internal/controller/kubecertagent/deployer.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+	for _, pod := range pods {
+		if pod.Status.Phase == corev1.PodRunning && (result == nil || newer(pod, result)) {
+			result = pod
+		}
+	}
+	return result
+}
+
+func int32ptr(i int32) *int32 { return &i }
+func int64ptr(i int64) *int64 { return &i }
```
Comment (enj) [inline internal/controller/kubecertagent/deployer.go:172]: Kube util code should have helpers for all of these.

File: internal/controller/kubecertagent/deployer.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+					Tolerations:                  controllerManagerPod.Spec.Tolerations,
+					// We need to run the agent pod as root since the file permissions
+					// on the cluster keypair usually restricts access to only root.
+					SecurityContext: &corev1.PodSecurityContext{
+						RunAsUser:  int64ptr(0),
+						RunAsGroup: int64ptr(0),
+					},
+				},
+			},
+			MinReadySeconds: 10,
```
Comment (enj) [inline internal/controller/kubecertagent/deployer.go:233]: Comment?

File: internal/controller/kubecertagent/deployer.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+		},
+	}
+}
+
+func mergeObjectMeta(existing metav1.ObjectMeta, expected metav1.ObjectMeta) metav1.ObjectMeta {
```
Comment (enj) [inline internal/controller/kubecertagent/deployer.go:238]: :thinking: I think there is a helper somewhere for this.

File: internal/controller/kubecertagent/execer.go [hunk 5]
```diff
@@ -98,7 +98,8 @@ func (c *execerController) Sync(ctx controllerlib.Context) error {
 		return nil
 	}
 
-	certPath, keyPath := c.getKeypairFilePaths(maybeAgentPod)
+	certPath := getPodEnv(maybeAgentPod, "CERT_PATH")
```
Comment (enj) [inline internal/controller/kubecertagent/execer.go:101]: Could you see if doing `cat ${env_var}` in the exec command works?

**Review by enj — COMMENTED** [review]
Minor stuff.  I have not looked at the tests.

File: internal/controller/kubecertagent/execer.go [hunk 6]
```diff
@@ -98,7 +98,8 @@ func (c *execerController) Sync(ctx controllerlib.Context) error {
 		return nil
```
Comment (enj) [inline internal/controller/kubecertagent/execer.go:98]: Reminder that this informer is incorrect (sync func is not singleton):

```go
		withInformer(
			configMapInformer,
			pinnipedcontroller.NameAndNamespaceExactMatchFilterFactory(clusterInfoName, ClusterInfoNamespace),
			controllerlib.InformerOption{},
		),
```

File: internal/controller/kubecertagent/deployer.go (hunk 1 above)
Comment (mattmoyer) [reply internal/controller/kubecertagent/deployer.go:146]: Doh, refactored this from some code that was reading directly from the client and missed this. Luckily our unit tests in CI caught it too 🎉

File: internal/controller/kubecertagent/deployer.go (hunk 2 above)
Comment (mattmoyer) [reply internal/controller/kubecertagent/deployer.go:172]: Ah yeah, in `k8s.io/utils/pointer`. I thought I remembered them existing but it was so easy to just rewrite them. Fixed.

File: internal/controller/kubecertagent/deployer.go (hunk 4 above)
Comment (mattmoyer) [reply internal/controller/kubecertagent/deployer.go:238]: I found the OpenShift helper but not anything in core kubernetes: https://github.com/openshift/library-go/blob/7a65fdb398e28782ee1650959a5e0419121e97ae/pkg/operator/resource/resourcemerge/object_merger.go#L12

File: internal/controller/kubecertagent/execer.go (hunk 5 above)
Comment (mattmoyer) [reply internal/controller/kubecertagent/execer.go:101]: I tried this and it does work. We can also use a single `sh -c 'cat ${CERT_PATH}; echo; echo; cat ${KEY_PATH}'` command and easily do this in a single exec call. I think I'm going to leave this as-is until the switch to distroless (#448) because that will require some similar changes anyway.

File: internal/controller/kubecertagent/legacypodcleaner.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+type legacyPodCleanerController struct {
+	cfg       AgentConfig
+	client    *kubeclient.Client
+	agentPods corev1informers.PodInformer
+	log       logr.Logger
+}
+
+// Sync implements controllerlib.Syncer.
+func (c *legacyPodCleanerController) Sync(ctx controllerlib.Context) error {
```
Comment (mattmoyer) [inline internal/controller/kubecertagent/legacypodcleaner.go:68]: TODO: inline this with the controllerlib closure helper.

File: test/integration/concierge_kubecertagent_test.go [hunk 8]
```diff
@@ -6,169 +6,85 @@ package integration
...
-	agentPodsReconciled := func() bool {
-		var currentAgentPods *corev1.PodList
-		currentAgentPods, err = kubeClient.CoreV1().Pods(env.ConciergeNamespace).List(ctx, metav1.ListOptions{
-			LabelSelector: kubeCertAgentLabelSelector,
+	// Expect there to be at least on healthy kube-cert-agent pod on this cluster.
+	library.RequireEventuallyWithoutError(t, func() (bool, error) {
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+		agentPods, err := kubeClient.CoreV1().Pods(env.ConciergeNamespace).List(ctx, metav1.ListOptions{
+			LabelSelector: "kube-cert-agent.pinniped.dev=true",
```
Comment (mattmoyer) [inline test/integration/concierge_kubecertagent_test.go:31]: TODO: this needs an update to v2.

**Review by enj — APPROVED** [review]
